File: 30. Serialize deserialize.cpp
#include <bits/stdc++.h>
using namespace std;
struct Node
{
    int data;
    Node* left;
    Node* right;
    Node(int d) : data(d), left(nullptr), right(nullptr) {}
};
void serialize(Node* root, ostream& out)
{
    if (!root)
    {
        out << "# ";
        return;
    }
    out << root->data << " ";
    serialize(root->left, out);
    serialize(root->right, out);
}
Node* deserialize(istringstream& in)
{
    string val;
    if (!(in >> val))
        return nullptr;
    if (val == "#")
        return nullptr;
    Node* root = new Node(stoi(val));
    root->left = deserialize(in);
    root->right = deserialize(in);
    return root;
}
int main()
{
    Node* root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    ostringstream out;
    serialize(root, out);
    string s = out.str();
    cout << s << endl;
    istringstream in(s);
    Node* r = deserialize(in);
    return 0;
}

File: 30. Serialize deserialize.py
class Node:
    def __init__(self,data): self.data=data; self.left=None; self.right=None

"""
Task: Serialize and deserialize a binary tree
Time Complexity: O(n)
Space Complexity: O(n)
"""

def serialize(root):
    res=[]
    def helper(node):
        if not node:
            res.append('#')
            return
        res.append(str(node.data))
        helper(node.left); helper(node.right)
    helper(root)
    return ' '.join(res)

def deserialize(s):
    vals=iter(s.split())
    def helper():
        val=next(vals)
        if val=='#': return None
        node=Node(int(val))
        node.left=helper()
        node.right=helper()
        return node
    return helper()

if __name__=='__main__':
    root=Node(1); root.left=Node(2); root.right=Node(3); root.left.left=Node(4)
    s=serialize(root)
    print('Serialized:', s)
    r=deserialize(s)
    print('Deserialized root:', r.data if r else None)
